index.js

```diff
@@ -644,13 +637,44 @@
   };
 
   //  Array$prototype$traverse :: Applicative f => Array a ~> (a -> f b, c -> f c) -> f (Array b)
-  var Array$prototype$traverse = function(f, of) {
-    var applicative = of([]);
-    for (var idx = this.length - 1; idx >= 0; idx -= 1) {
-      applicative = ap(map(prepend, f(this[idx])), applicative);
-    }
-    return applicative;
-  };
+  var Array$prototype$traverse = (function() {
+    var array1 = function(x) {
+      return [x];
+    };
+    var array2 = function(x) {
+      return function(y) {
+        return [x, y];
+      };
+    };
+    var array3 = function(x) {
+      return function(y) {
+        return function(z) {
+          return [x, y, z];
+        };
+      };
+    };
+    var concat2 = function(xs) {
+      return function(ys) {
+        return xs.concat(ys);
+      };
+    };
+    return function(f, of) {
+      var go = function(bottom, top) {
+        switch (top - bottom) {
+          case 0: return of([]);
+          case 1: return map(array1, f(this[bottom]));
+          case 2: return lift2(array2, f(this[bottom]), f(this[bottom + 1]));
+          case 3: return lift3(array3, f(this[bottom]), f(this[bottom + 1]),
+                               f(this[bottom + 2]));
+          default:
+            var pivot = bottom + Math.floor((top - bottom) / 4) * 2;
+            return lift2(concat2, go.call(this, bottom, pivot),
+                         go.call(this, pivot, top));
+        }
+      };
+      return go.call(this, 0, this.length);
```

Comment:
Is there a reason to provide the array as an implicit argument? Could we use `go(this, 0, this.length)` instead?